src/node_dir.cc

```diff
@@ -160,7 +160,66 @@ void DirHandle::Close(const FunctionCallbackInfo<Value>& args) {
   }
 }
 
-void AfterDirReadSingle(uv_fs_t* req) {
+static MaybeLocal<Array> DirentListToArray(
+    Environment* env,
+    uv_dirent_t* ents,
+    int num,
+    enum encoding encoding,
+    Local<Value>* err_out) {
+  MaybeStackBuffer<Local<Value>, 96> entries(num * 3);
+
+  // Return an array of all read filenames.
+  int j = 0;
+  for (int i = 0; i < num; i++) {
+    Local<Value> filename;
+    Local<Value> error;
+    if (!StringBytes::Encode(env->isolate(),
+                             ents[i].name,
+                             encoding,
+                             &error).ToLocal(&filename)) {
+      *err_out = error;
+      return MaybeLocal<Array>();
+    }
+
+    entries[j++] = filename;
+
+    // We cache the returned values in JS. However, the API allows multiple
+    // read() calls with different `encoding` options; while this may be
+    // (extremely?) uncommon, it means that there is a possibility that the
+    // `filename` string above that was decoded using the original encoding
+    // may not be sufficient for providing the underlying byte sequence in
+    // another encoding.
+    // In particular, when using the ASCII encoding, UTF-8 with invalid data
+    // or UCS2 with odd byte count, the decoding process may not be reversible.
+    // In those cases, we store an additional Buffer instance containing the
+    // original underlying memory.
+    // (We cannot call indexOf() on strings from C++, so for UTF-8, we check
+    // whether the data contains only codepoints <= 255, in which case we
+    // can be sure that the data is valid. Ideally, we’d be searching for
+    // U+FFFD '�' explictly instead. Also, using UCS2 here is nearly useless
+    // anyway because libuv does not know the length of the string.)
+    // We skip this step for the first item because that is never going to be
+    // read using another encoding.
+    if (i > 0 &&
+        ((encoding == ASCII) ||
+         (encoding == UTF8 && !filename.As<String>()->ContainsOnlyOneByte()) ||
+         (encoding == UCS2 && strlen(ents[i].name) % 2 != 0))) {
+      Local<Object> fallback_buf;
+      if (!Buffer::Copy(env,
+                        ents[i].name,
+                        strlen(ents[i].name)).ToLocal(&fallback_buf)) {
```

Comment:
nit: can store the length in a var eventually see above.

Reply:
Done